src/operator/timestamp.ts

```diff
@@ -0,0 +1,45 @@
+import {Operator} from '../Operator';
+import {Observable} from '../Observable';
+import {Subscriber} from '../Subscriber';
+import {Scheduler} from '../Scheduler';
+import {async} from '../scheduler/async';
+
+/**
+ * @param scheduler
+ * @return {Observable<TimeStamp<any>>|WebSocketSubject<T>|Observable<T>}
+ * @method timestamp
+ * @owner Observable
+ */
+export function timestamp<T>(scheduler: Scheduler = async): Observable<TimeStamp<T>> {
+  return this.lift(new TimeStampOperator(scheduler));
+}
+
+export interface TimeStampSignature<T> {
+  (scheduler?: Scheduler): Observable<TimeStamp<T>>;
+}
+
+export class TimeStamp<T> {
+  constructor(public value: T, public timestamp: number) {
+  }
+};
+
+class TimeStampOperator<T> implements Operator<T, TimeStamp<T>> {
+  constructor(private scheduler: Scheduler) {
+  }
+
+  call(observer: Subscriber<TimeStamp<T>>): Subscriber<T> {
+    return new TimeStampSubscriber(observer, this.scheduler);
+  }
+}
+
+class TimeStampSubscriber<T> extends Subscriber<T> {
+  constructor(destination: Subscriber<TimeStamp<T>>, private scheduler: Scheduler) {
+    super(destination);
+  }
+
+  protected _next(value: T): void {
+    const now = this.scheduler.now();

```

Comment:
For now, I'd suggest to go as is. First, it has default scheduler set behavior when not provided. Secondly, it's case someone puts null intentionally - maybe good to have later but it doesn't feel to should be corrected internally. What do you think?